### PR TITLE
sql/schemachanger: address bugs that can lead to hanging declarative schema change jobs

### PR DESCRIPTION
**Pull request by fqazi**
This PR addresses two separate issues that lead to similar problems, where we can hang because the schema changer state is no longer constructible.

1. There is a small window in the declarative schema changer in drop scenarios where if the job succeeds, but the status cannot be updated we may re-attempt the run phase. When we do this we will try to get the declarative schema changer state based on the descriptor list, if any descriptor is dropped then we will no longer be able to construct the state. This PR will check the status to see if all stages are completed and treat this scenario as a no-op.
2. When running a DROP OWNED BY (declarative) and DROP DATABASE (legacy) concurrently we could run into a hang since the DROP DATABASE would not wait for the declarative schema change to complete properly. This PR will add concurrent schema change checks for DROP DATABASE/ DROP SCHEMA/ DROP TYPE.

### CODE REVIEW

**Comment by ajwerner** [comment]
Great find!

File: pkg/sql/drop_database.go [hunk 1]
```diff
@@ -128,6 +129,12 @@ func (n *dropDatabaseNode) startExec(params runParams) error {
 	ctx := params.ctx
 	p := params.p
 
+	// Exit early with an error if the schema is undergoing a declarative schema
+	// change.
+	if n.dbDesc.GetDeclarativeSchemaChangerState() != nil {
```
Comment (postamar) [inline pkg/sql/drop_database.go:134]: Why not `if n.dbDesc.HasConcurrentSchemaChanges() {` instead? Same comment applies for other legacy DROPs.

**Review by postamar — COMMENTED** [review]
Just a couple of comments related to the implementation of these changes, otherwise the changes themselves seem sound to me.

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88923)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar)

___
*[`pkg/sql/drop_database.go` line 134 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/88923#-ND3pbXooMrQYFbS-9cb-r2-134:-ND8qwrk3KhP3f6mCOil:b-ear8b9) ([raw file](https://github.com/cockroachdb/cockroach/blob/82422df1162b97aa867cf5694bd3c7ffdfdef839/pkg/sql/drop_database.go#L134)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Why not `if n.dbDesc.HasConcurrentSchemaChanges() {` instead? Same comment applies for other legacy DROPs.
</blockquote></details>

Done.

Yes, updated both legacy and current ones to use this interface.
___
*[`pkg/sql/schemachanger/scjob/job.go` line 105 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88923#-ND3pbXooMrQYFbS-9cY-r1-105:-ND8r4bJ-DqucmAJnH5r:b-8qzhdp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9be032b7a8998a909266c9d79bedbedecc2a63a/pkg/sql/schemachanger/scjob/job.go#L105)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Rather than relying on this job status string, couldn't we rely on the job's descriptor IDs instead? If possible I'd rather not depend on a user-facing string's contents to determine job doneness. It seems to me that "all stages completed" will be reported if and only if the job's descriptor ID set is empty.
</blockquote></details>

Done.

Yes, a cleaner / nicer solution is just to update the descriptor ID set. We already did this during job progress, so I modified this code to clear the entire set once the last stage is executed.

**Review by ajwerner — APPROVED** [review]
:lgtm:

> Reviewed 16 of 183 files at r3, 6 of 8 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88923#-:-ND9XW-a1EW-kb8DlqFa:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @postamar)

**Comment by ajwerner** [comment]
Hmm, seems like there is some fallout

**Comment by fqazi** [comment]
@ajwerner TFTR!

Ah, the legacy schema changer is not as disciplined about writing job ID's. So, I reverted to the old logic as a helper function with an explanation. I'll make sure CI is happy before letting bors loose.

**Comment by ajwerner** [comment]
It's than and it's that the legacy schema changer is happy to, say, drop a table that has ongoing schema changes. I forgot that that was true.

**Comment by fqazi** [comment]
@ajwerner Begs the question if we should allow the same behavior for declarative

**Comment by fqazi** [comment]
blathers backport release-22.2.0

**Comment by ajwerner** [comment]
@fqazi I think this got backported into 22.2.0 but not 22.2.x. 

blather backport 22.2

**Pull request merged**